File: VB/DxSample.Module/DatabaseUpdate/Updater.vb
Imports System
Imports System.IO
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.Updating
Imports DxSample.Module.BusinessObjects

Namespace DxSample.Module.DatabaseUpdate
	Public Class Updater
		Inherits ModuleUpdater

		Public Sub New(ByVal objectSpace As IObjectSpace, ByVal currentDBVersion As Version)
			MyBase.New(objectSpace, currentDBVersion)
		End Sub

		Public Overrides Sub UpdateDatabaseAfterUpdateSchema()
			MyBase.UpdateDatabaseAfterUpdateSchema()
			Me.CreateCustomers()
			Me.CreateOrders()
		End Sub

		Private Sub CreateObjects(Of TEntity)(ByVal resourceName As String, ByVal updateEntity As Action(Of TEntity, String()))
			Dim objectsCnt As Integer = Me.ObjectSpace.GetObjectsCount(GetType(TEntity), Nothing)
			If objectsCnt > 0 Then
				Return
			End If
			Dim stream As Stream = GetType(TEntity).Assembly.GetManifestResourceStream(resourceName)
			Using reader As TextReader = New StreamReader(stream)
				Dim line As String = String.Empty
                line = reader.ReadLine()
                Do While Not String.IsNullOrEmpty(line)
                    Dim data() As String = line.Split(";"c)
                    Dim entity As TEntity = Me.ObjectSpace.CreateObject(Of TEntity)()
                    updateEntity(entity, data)
                    line = reader.ReadLine()
                Loop
			End Using
			Me.ObjectSpace.CommitChanges()
		End Sub

		Private Sub UpdateCustomer(ByVal customer As Customer, ByVal data() As String)
			customer.FirstName = data(0)
			customer.LastName = data(1)
			customer.Country = data(2)
		End Sub

		Private Sub UpdateOrder(ByVal order As Order, ByVal customer As Customer, ByVal data() As String)
			order.Customer = customer
			order.OrderDate = Date.Parse(data(0))
			order.UnitPrice = Decimal.Parse(data(1))
		End Sub

		Private Sub CreateCustomers()
			Me.CreateObjects(Of Customer)("DxSample.Module.Resources.Customers.csv", AddressOf Me.UpdateCustomer)
		End Sub

		Private Sub CreateOrders()
			Dim customers = Me.ObjectSpace.GetObjects(Of Customer)()
			Dim randomizer As New Random(Date.Now.Millisecond)
			Me.CreateObjects(Of Order)("DxSample.Module.Resources.Orders.csv", Sub(order, data) Me.UpdateOrder(order, customers(randomizer.Next(customers.Count)), data))
		End Sub
	End Class
End Namespace

File: CS/WebChart/WebChart.Module/DatabaseUpdate/Updater.cs
using System;
using System.IO;
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.Updating;
using DxSample.Module.BusinessObjects;

namespace DxSample.Module.DatabaseUpdate {
    public class Updater : ModuleUpdater {
        public Updater(IObjectSpace objectSpace, Version currentDBVersion) :
            base(objectSpace, currentDBVersion) {
        }

        public override void UpdateDatabaseAfterUpdateSchema() {
            base.UpdateDatabaseAfterUpdateSchema();
            this.CreateCustomers();
            this.CreateOrders();
        }

        private void CreateObjects<TEntity>(string resourceName, Action<TEntity, string[]> updateEntity) {
            int objectsCnt = this.ObjectSpace.GetObjectsCount(typeof(TEntity), null);
            if(objectsCnt > 0) return;
            Stream stream = typeof(TEntity).Assembly.GetManifestResourceStream(resourceName);
            using(TextReader reader = new StreamReader(stream)) {
                string line = string.Empty;
                while(!string.IsNullOrEmpty(line = reader.ReadLine())) {
                    string[] data = line.Split(';');
                    TEntity entity = this.ObjectSpace.CreateObject<TEntity>();
                    updateEntity(entity, data);
                }
            }
            this.ObjectSpace.CommitChanges();
        }

        private void UpdateCustomer(Customer customer, string[] data) {
            customer.FirstName = data[0];
            customer.LastName = data[1];
            customer.Country = data[2];
        }

        private void UpdateOrder(Order order, Customer customer, string[] data) {
            order.Customer = customer;
            order.OrderDate = DateTime.Parse(data[0]);
            order.UnitPrice = decimal.Parse(data[1]);
        }

        private void CreateCustomers() {
            this.CreateObjects<Customer>("WebChart.Module.Resources.Customers.csv", this.UpdateCustomer);
        }

        private void CreateOrders() {
            var customers = this.ObjectSpace.GetObjects<Customer>();
            Random randomizer = new Random(DateTime.Now.Millisecond);
            this.CreateObjects<Order>("WebChart.Module.Resources.Orders.csv",
                (order, data) => this.UpdateOrder(order, customers[randomizer.Next(customers.Count)], data));
        }
    }
}
